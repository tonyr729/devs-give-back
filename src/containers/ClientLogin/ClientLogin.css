.client-background {
  background-color: #607d8b;
  width: 100vw;
  height: 100vh; }

.login-message {
  font-size: 5rem;
  padding-top: 12rem; }

.login-button {
  border: none;
  width: 12rem;
  height: 2.9rem;
  background-repeat: no-repeat;
  background-image: url("../../images/btn_google_light_normal_ios.svg");
  padding-left: 3rem;
  font-size: 1rem;
  color: #6d6c6c;
  margin: 4rem auto;
  display: block; }
  .login-button:hover {
    color: #f8f8ff;
    background-image: url("../../images/btn_google_light_focus_ios.svg");
    background-color: #448aff; }
  .login-button:active {
    color: #f8f8ff;
    background-image: url("../../images/btn_google_light_focus_ios.svg");
    background-color: #24498a; }
  .login-button:focus {
    outline: none; }
