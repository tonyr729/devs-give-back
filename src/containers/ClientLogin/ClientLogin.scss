//COLORS
$grey: #4f4f4f;
$white: #ffffff;
$green: #39bda7;
$blue: #03a9f4;
$transparent: rgba(255, 255, 255, 0);
$black-05: rgba(0, 0, 0, 0.05);

.client-background {
  background-color: $white;
  width: 100vw;
  height: 100vh;
}

.client-login-message {
  font-size: 5rem;
  padding-top: 12rem;
  color:$grey;
}

.login-button {
  border: none;
  border-radius: .3rem;
  width: 12rem;
  height: 2.9rem;
  background-repeat: no-repeat;
  background-image: url('../../images/btn_google_light_normal_ios.svg');
  padding-left: 3rem;
  font-size: 1rem;
  color: #6d6c6c;
  margin: 4rem auto;
  display: block;
  background-color: $blue;
  color: $white;

  &:hover {
    color: $white;
    background-image: url('../../images/btn_google_light_focus_ios.svg');
    background-color: #597cb8;
  }
  &:active {
    color: #f8f8ff;
    background-image: url('../../images/btn_google_light_focus_ios.svg');
    background-color: #24498a;
  }
  &:focus {
    outline: none;
  }
}