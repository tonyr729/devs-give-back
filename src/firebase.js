import firebase from 'firebase';

const config = {
  apiKey: "AIzaSyAVk4m35HpmPoi0Rr1K495lVyfskmVDiD4",
  authDomain: "devsgiveback.firebaseapp.com",
  databaseURL: "https://devsgiveback.firebaseio.com",
  projectId: "devsgiveback",
  storageBucket: "devsgiveback.appspot.com",
  messagingSenderId: "171321681886"
};
firebase.initializeApp(config);
export default firebase;