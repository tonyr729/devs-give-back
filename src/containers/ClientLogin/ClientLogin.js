import React, { Component } from 'react';
import firebase from '../../firebase/firebase';
import './ClientLogin.css';
import { connect } from 'react-redux';
import { signInClient, clientError } from '../../actions/actions';
import DataCleaner from '../../helpers/DataCleaner';

const cleaner = new DataCleaner();

export class ClientLogin extends Component {
  
  googleLogin = async () => {
    try {
      const provider = new firebase.auth.GoogleAuthProvider();
      const response = await firebase.auth().signInWithPopup(provider);
      const cleanUser = cleaner.cleanClientLogin(response.user)
      this.handleUser(cleanUser)
      this.props.history.push("/problem-title");
    } catch (error) {
      const errorCode = error.code;
      const errorMessage = error.message
      const email = error.email;
      const credential = error.credential;
      const errorInfo = { errorCode, errorMessage, email, credential};
      this.props.clientError(errorInfo);
      this.props.history.push("/error-page")
    }
  };
  
  handleUser = (user) => {
    this.props.signInClient(user);
    firebase.database().ref('users/' + user.id).set({
      username: user.name,
      picture: user.photoURL
    });
  }
  
  render() {
    return (
      <div>
        <p className='login-message'>Please login</p>
        <button className='login-button'onClick={()=> this.googleLogin()}>Login with Google</button>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  user: state.client,
  clientError: state.clientError
})

const mapDispatchToProps = (dispatch) => ({
  signInClient: (user) => dispatch(signInClient(user)),
  clientError: (error) => dispatch(clientError(error))
});

export default connect(mapStateToProps, mapDispatchToProps)(ClientLogin);
