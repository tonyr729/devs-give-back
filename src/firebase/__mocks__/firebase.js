import * as firebase from 'firebase';

const onAuthStateChanged = jest.fn();

const getRedirectResult = jest.fn(() => {
  return Promise.resolve({
    user: {
      displayName: 'redirectResultTestDisplayName',
      email: 'redirectTest@test.com',
      emailVerified: true
    }
  })
})

const sendEmailVerification = jest.fn(() => {
  return Promise.resolve('result of sendEmailVerification');
})

const sendPasswordResetEmail = jest.fn(() => Promise.resolve());

const createUserWithEmailAndPassword = jest.fn(() => {
  return Promise.resolve('result of createUserWithEmailAndPassword');
})

const signInWithEmailAndPassword = jest.fn(() => {
  return Promise.resolve('result of signInWithEmailAndPassword');
})

const signInWithPopup = jest.fn(() => {
  return Promise.resolve('result of signInWithPopup');
})

const initializeApp = jest
  .spyOn(firebase, 'initializeApp')
  .mockImplementation(() => {
    return {
      auth: () => {
        return {
          createUserWithEmailAndPassword,
          signInWithEmailAndPassword,
          currentUser: {
            sendEmailVerification
          }
        }
      }
    }
  })

jest.spyOn(firebase, 'auth').mockImplementation(() => {
  return {
    onAuthStateChanged,
    currentUser: {
      displayName: 'testDisplayName',
      email: 'test@test.com',
      emailVerified: true
    },
    signInWithPopup
  };
});

firebase.auth.GithubAuthProvider = jest.fn(() => {})
firebase.auth.GoogleAuthProvider = jest.fn(() => {})

export default firebase;