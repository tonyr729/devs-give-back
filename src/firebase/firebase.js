import firebase from 'firebase';
import { firebaseKey } from './apiKeys';

const config = {
  apiKey: firebaseKey,
  authDomain: "devsgiveback.firebaseapp.com",
  databaseURL: "https://devsgiveback.firebaseio.com",
  projectId: "devsgiveback",
  storageBucket: "devsgiveback.appspot.com",
  messagingSenderId: "171321681886"
};
firebase.initializeApp(config);

export default firebase;