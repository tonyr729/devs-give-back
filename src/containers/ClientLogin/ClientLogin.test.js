import { shallow, mount } from 'enzyme';
import React from 'react';
import { ClientLogin } from './ClientLogin';
import firebase from '../../firebase/firebase'
import { Redirect } from 'react-router-dom';
jest.mock('../../firebase/firebase');
jest.mock('../../helpers/DataCleaner.js', () => {
  return jest.fn().mockImplementation(() => {
    return {
      cleanClientLogin: jest.fn().mockImplementation(() => Promise.resolve({user: 'name'})),
      cleanError: jest.fn().mockImplementation(() => Promise.resolve({error: 'error'}))
    };
  });
});


describe('ClientLogin', () => {
  let mockProps;
  let clientLogin

  beforeEach(() => {
    mockProps = {
      client: { },
      clientError: jest.fn(),
      signInClient: jest.fn(),
      history: { push: jest.fn() }
    };
    
    clientLogin = shallow(<ClientLogin {...mockProps} />);
  });

  it('should match snapshot', () => {
    expect(clientLogin).toMatchSnapshot();
  });

  describe('googleLogin', () => {
    let clientLogin;
    let mockProps;

    beforeEach(() => {
      mockProps = {
        client: { },
        clientError: jest.fn(),
        signInClient: jest.fn(),
        history: { push: jest.fn() }
      };

      clientLogin = shallow(<ClientLogin {...mockProps} />);
    });

    it('should call the expected methods of firebase', () => {
      clientLogin.find('button').simulate('click');

      expect(firebase.auth().signInWithPopup).toHaveBeenCalled()
      expect(firebase.auth.GoogleAuthProvider).toHaveBeenCalled()
    });
  });

  describe('logInCheck', () => {
    let mockProps;
    let clientLogin

    beforeEach(() => {
      mockProps = {
        client: { },
        clientError: jest.fn(),
        signInClient: jest.fn(),
        history: { push: jest.fn() }
      };
      
      clientLogin = shallow(<ClientLogin {...mockProps} />);
    });
    
    it('should call redirect if client is not signed in', () => {
      
      const expected = clientLogin.instance().logInCheck({client: 'tony'})
      
      expect(expected).toEqual(<Redirect push={false} to='/problem-title'/>);
    });
  });
  
  describe('handleLogin', () => {
    let mockProps;
    let clientLogin
    
    beforeEach(() => {
      mockProps = {
        client: { },
        clientError: jest.fn(),
        signInClient: jest.fn(),
        history: { push: jest.fn() }
      };
      
      clientLogin = shallow(<ClientLogin {...mockProps} />);
      clientLogin.instance().api.googleLogin = jest.fn().mockImplementation(() => Promise.resolve({user: 'name'}));
    });
      
    
    it('should call the required functions when called', async () => {
      const instance = clientLogin.instance()
      
      await instance.handleLogin()
      
      expect(instance.api.googleLogin).toHaveBeenCalled()
      expect(instance.cleaner.cleanClientLogin).toHaveBeenCalled();
      expect(instance.props.signInClient).toHaveBeenCalled();
    });
  });


  describe('writeToDatabase', () => {
    let mockProps;
    let clientLogin;

    beforeEach(() => {
      mockProps = {
        client: { },
        clientError: jest.fn(),
        signInClient: jest.fn(),
        history: { push: jest.fn() }
      };
      
      clientLogin = shallow(<ClientLogin {...mockProps} />);
    });

    it('should send client to database', () => {
      clientLogin.instance().writeToDatabase(mockProps.client)


      expect(firebase.database().ref().set).toHaveBeenCalled();
    });
  });
});
