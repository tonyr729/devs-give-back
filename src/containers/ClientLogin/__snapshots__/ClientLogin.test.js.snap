// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientLogin should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ClientLogin />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="login-message"
>
          Please login
</p>,
        <button
          className="login-button"
          onClick={[Function]}
>
          Login with Google
</button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Please login",
          "className": "login-message",
        },
        "ref": null,
        "rendered": "Please login",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login with Google",
          "className": "login-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Login with Google",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="login-message"
>
            Please login
</p>,
          <button
            className="login-button"
            onClick={[Function]}
>
            Login with Google
</button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Please login",
            "className": "login-message",
          },
          "ref": null,
          "rendered": "Please login",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login with Google",
            "className": "login-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Login with Google",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
